lib/connect/ws.js

```diff
@@ -1,257 +1,265 @@
-    stream.emit('connect')
-  }
-
-  function onclose () {
-    stream.end()
-    stream.destroy()
-  }
-
-  function onerror (err) {
-    stream.destroy(err)
-  }
-
-  function onmessage (event) {
-    let data = event.data
-    if (data instanceof ArrayBuffer) data = Buffer.from(data)
-    else data = Buffer.from(data, 'utf8')
-    proxy.push(data)
-  }
-
-  // this is to be enabled only if objectMode is false
-  function writev (chunks, cb) {
-    const buffers = new Array(chunks.length)
-    for (let i = 0; i < chunks.length; i++) {
-      if (typeof chunks[i].chunk === 'string') {
-        buffers[i] = Buffer.from(chunks[i], 'utf8')
-      } else {
-        buffers[i] = chunks[i].chunk
-      }
-    }
-
-    this._write(Buffer.concat(buffers), 'binary', cb)
-  }
-
-  function socketWriteBrowser (chunk, enc, next) {
-    if (socket.bufferedAmount > bufferSize) {
-      // throttle data until buffered amount is reduced.
-      setTimeout(socketWriteBrowser, bufferTimeout, chunk, enc, next)
-    }
-
-    if (coerceToBuffer && typeof chunk === 'string') {
-      chunk = Buffer.from(chunk, 'utf8')
-    }
-
-    try {
-      socket.send(chunk)
-    } catch (err) {
-      return next(err)
-    }
-
-    next()
-  }
-
-  function socketEndBrowser (done) {
-    socket.close()
-    done()
-  }
-
-  // end methods for browserStreamBuilder
-
-  return stream
+function browserStreamBuilder(client, opts) {
+	debug('browserStreamBuilder')
+	let stream
+	const options = setDefaultBrowserOpts(opts)
+	// sets the maximum socket buffer size before throttling
+	const bufferSize = options.browserBufferSize || 1024 * 512
+
+	const bufferTimeout = opts.browserBufferTimeout || 1000
+
+	const coerceToBuffer = !opts.objectMode
+
+	const socket = createBrowserWebSocket(client, opts)
+
+	const proxy = buildProxy(opts, socketWriteBrowser, socketEndBrowser)
+
+	if (!opts.objectMode) {
+		proxy._writev = writev
+	}
+	proxy.on('close', () => {
+		socket.close()
+	})
+
+	const eventListenerSupport = typeof socket.addEventListener !== 'undefined'
+
+	// was already open when passed in
+	if (socket.readyState === socket.OPEN) {
+		stream = proxy
+	} else {
+		stream = duplexify(undefined, undefined, opts)
+		if (!opts.objectMode) {
+			stream._writev = writev
+		}
+
+		if (eventListenerSupport) {
+			socket.addEventListener('open', onopen)
+		} else {
+			socket.onopen = onopen
+		}
+	}
+
+	stream.socket = socket
+
+	if (eventListenerSupport) {
+		socket.addEventListener('close', onclose)
+		socket.addEventListener('error', onerror)
+		socket.addEventListener('message', onmessage)
+	} else {
+		socket.onclose = onclose
+		socket.onerror = onerror
+		socket.onmessage = onmessage
+	}
+
+	// methods for browserStreamBuilder
+
+	function buildProxy(pOptions, socketWrite, socketEnd) {
+		const _proxy = new Transform({
+			objectModeMode: pOptions.objectMode,
+		})
+
+		_proxy._write = socketWrite
+		_proxy._flush = socketEnd
+
+		return _proxy
+	}
+
+	function onopen() {
```

Comment:
nit: camelCase this
Might be out of scope for this PR
```suggestion
	function onOpen() {
```

Reply:
Done, anyway the reason this wasn't camelacase is to follow `socket.onopen` that is not.